crates/biome_graphql_parser/src/parser/definitions/union.rs

```diff
@@ -0,0 +1,139 @@
+use crate::parser::{
+    directive::DirectiveList,
+    is_at_name, parse_description,
+    parse_error::{expected_name, expected_named_type},
+    parse_name,
+    r#type::parse_named_type,
+    value::is_at_string,
+    GraphqlParser,
+};
+use biome_graphql_syntax::{
+    GraphqlSyntaxKind::{self, *},
+    T,
+};
+use biome_parser::{
+    parse_lists::{ParseNodeList, ParseSeparatedList},
+    parse_recovery::ParseRecovery,
+    parsed_syntax::ParsedSyntax,
+    prelude::ParsedSyntax::*,
+    token_source::TokenSource,
+    Parser,
+};
+
+use super::is_at_definition;
+
+#[inline]
+pub(crate) fn parse_union_type_definition(p: &mut GraphqlParser) -> ParsedSyntax {
+    if !is_at_union_type_definition(p) {
+        return Absent;
+    }
+    let m = p.start();
+
+    // description is optional
+    parse_description(p).ok();
+
+    p.bump(T![union]);
+
+    parse_name(p).or_add_diagnostic(p, expected_name);
+
+    DirectiveList.parse_list(p);
+
+    // union member types are optional
+    parse_union_member_types(p).ok();
+
+    Present(m.complete(p, GRAPHQL_UNION_TYPE_DEFINITION))
+}
+
+#[inline]
+fn parse_union_member_types(p: &mut GraphqlParser) -> ParsedSyntax {
+    if !is_at_union_member_types(p) {
+        return Absent;
+    }
+    let m = p.start();
+    p.expect(T![=]);
+
+    if p.at(T![|]) {
+        p.bump(T![|]);
+    }
+
+    let position = p.source().position();
+    UnionMemberTypeList.parse_list(p);
+
+    // has not progressed, meaning no union member types were parsed
+    if position == p.source().position() {
+        p.error(expected_named_type(p, p.cur_range()));
+    }
```

Comment:
Is there a better way to check if `UnionMemberTypeList` was empty?

Reply:
Internally, ParsedLists, check the progression of the parser and they fail if that parser hasn't advanced.

Check the source code to make sure

Reply:
Well, this is exactly the approach used by parse_list. However, parse_list uses ParseProgress, which has a private constructor and panic if the parser doesn't progress, so I just simplified it to this.